src/CompoundRegion.cc

```diff
@@ -0,0 +1,196 @@
+    std::vector<std::uint8_t> buffer;
+    buffer.push_back(tc);
+    auto buffer1 = getOperand(0).encode();
+    encodeU64(buffer1.size(), buffer);
+    buffer.insert(buffer.end(), buffer1.begin(), buffer1.end());
+    auto buffer2 = getOperand(1).encode();
+    encodeU64(buffer2.size(), buffer);
+    buffer.insert(buffer.end(), buffer2.begin(), buffer2.end());
+    return buffer;
+}
+
+std::array<std::unique_ptr<Region>, 2> CompoundRegion::_decode(
+    std::uint8_t tc, std::uint8_t const *buffer, std::size_t nBytes) {
+    std::uint8_t const *end = buffer + nBytes;
+    if (nBytes == 0) {
+        throw std::runtime_error("Encoded CompoundRegion is truncated.");
+    }
+    if (buffer[0] != tc) {
+        throw std::runtime_error("Byte string is not an encoded CompoundRegion.");
+    }
+    ++buffer;
+    std::array<std::unique_ptr<Region>, 2> result;
+    std::uint64_t nBytes1 = consumeDecodeU64(buffer, end);
+    result[0] = Region::decode(buffer, nBytes1);
+    buffer += nBytes1;
+    std::uint64_t nBytes2 = consumeDecodeU64(buffer, end);
+    result[1] = Region::decode(buffer, nBytes2);
+    buffer += nBytes2;
+    if (buffer != end) {
+        throw std::runtime_error("Encoded CompoundRegion is has unexpected additional bytes.");
+    }
+    return result;
+}
+
+std::unique_ptr<CompoundRegion> CompoundRegion::decode(uint8_t const *buffer, size_t n) {
+    if (n == 0) {
+        throw std::runtime_error("Encoded CompoundRegion is truncated.");
+    }
+    switch (buffer[0]) {
+        case UnionRegion::TYPE_CODE:
+            return UnionRegion::decode(buffer, n);
+        case IntersectionRegion::TYPE_CODE:
+            return IntersectionRegion::decode(buffer, n);
+        default:
+            throw std::runtime_error("Byte string is not an encoded CompoundRegion.");
+    }
+}
+
+Box UnionRegion::getBoundingBox() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingBox(); });
+}
+
+Box3d UnionRegion::getBoundingBox3d() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingBox3d(); });
+}
+
+Circle UnionRegion::getBoundingCircle() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingCircle(); });
+}
+
+bool UnionRegion::contains(UnitVector3d const &v) const {
+    return getOperand(0).contains(v) || getOperand(1).contains(v);
+}
+
+Relationship UnionRegion::relate(Region const &rhs) const {
+    auto r1 = getOperand(0).relate(rhs);
+    auto r2 = getOperand(1).relate(rhs);
+    return
+        // Both operands must be disjoint with the given region for the union
+        // to be disjoint with it.
+        ((r1 & DISJOINT) & (r2 & DISJOINT))
+        // Both operands must be within the given region for the union to be
+        // within it.
+        | ((r1 & WITHIN) & (r2 & WITHIN))
+        // If either operand contains the given region, the union contains it.
+        | ((r1 & CONTAINS) | (r2 & CONTAINS));
```

Comment:
What about cases where neither r1 nor r2 individually contain rhs, but the union of r1 and r2 does (for example, two overlapping circles that together contain an ellipse not entirely contained by either one individually?)

Reply:
You're absolute right about the behavior, but we're saved by the limited guarantees made by the `Region` interface on the behavior; from the docs on `Region::relate`:
```
    /// Said another way: if the CONTAINS, WITHIN or DISJOINT bit is set, then
    /// the corresponding spatial relationship between the two regions holds
    /// conclusively. If it is not set, the relationship may or may not
    /// hold.
    ///
    /// These semantics allow for conservative relationship computations. In
    /// particular, a Region may choose to implement `relate` by replacing
    /// itself and/or the argument with a simplified bounding region.
```
It'd be nice to do better, but I think we'd at least need way for all `Region` classes to be able to compute some common set of shapes they bound (rather than are bounded by), and that's a lot of hard math for still not being able to formally guarantee any more than we do.